src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -349,44 +347,46 @@ public static unsafe bool TryFormatDecimal(decimal value, ReadOnlySpan<char> for
             return sb.TryCopyTo(destination, out charsWritten);
         }
 
-        internal static unsafe void DecimalToNumber(ref decimal d, ref NumberBuffer number)
+        internal static void DecimalToNumber(ref decimal d, ref NumberBuffer number)
         {
-            byte* buffer = number.GetDigitsPointer();
+            Span<byte> buffer = number.Digits;
             number.DigitsCount = DecimalPrecision;
             number.IsNegative = d.IsNegative;
 
-            byte* p = buffer + DecimalPrecision;
+            int p = DecimalPrecision;
             while ((d.Mid | d.High) != 0)
             {
-                p = UInt32ToDecChars(p, decimal.DecDivMod1E9(ref d), 9);
+                p = UInt32ToDecChars(buffer, p, decimal.DecDivMod1E9(ref d), 9);
             }
-            p = UInt32ToDecChars(p, d.Low, 0);
+            p = UInt32ToDecChars(buffer, p, d.Low, 0);
 
-            int i = (int)((buffer + DecimalPrecision) - p);
+            int i = DecimalPrecision - p;
 
             number.DigitsCount = i;
             number.Scale = i - d.Scale;
 
-            byte* dst = number.GetDigitsPointer();
+            int dst = 0;
             while (--i >= 0)
             {
-                *dst++ = *p++;
+                buffer[dst++] = buffer[p++];
             }
-            *dst = (byte)('\0');
+            buffer[dst] = (byte)('\0');
 
             number.CheckConsistency();
         }
 
         public static string FormatDouble(double value, string? format, NumberFormatInfo info)
         {
+            Span<byte> pDigits = stackalloc byte[DoubleNumberBufferLength];
```

Comment:
Why does this need to be passed in now, rather than keeping it where the stackalloc was previously?

Reply:
pDigits and sb needs to be at the same method, or compiler generates an error when calling NumberToString, I believe it is because NumberToString could theoretically assign pDigits to sb and escape pDigits beyond its stack frame:

![image](https://user-images.githubusercontent.com/1336227/103934346-94e94780-5103-11eb-9b71-c99362595b0b.png)